Bug 1331412 - Migrate Shared to Swift 3

sleroux opened the pull request:
This PR has the auto-migration as the first commit and all of the manual fixes broken down into the remaining parts. Some high level highlights:

* Alamofire.Manager is now known as SessionManager
* XCGLogger API has changed quite a bit but isn't too bad
* Changes to UnsafePointer/UnsafeMutablePointer are a bit unfamiliar. I tried to follow the migration guide section (here)[https://swift.org/migration-guide/se-0107-migrate.html]
* Various URL changes - mostly `path` now being non-optional
* json.swift is being a pain now that `JSONSerialization` returns `Any` type instead of `AnyObject`. I took the simplest approach here and make everything into `AnyObject`. The two types are bridged so it shouldn't be a problem. Just really annoying/messy. We should replace this eventually.
* Swift 3 introduces precedence groups instead of using magic numbers. IIRC monadic bind/do are higher than multiplication in Haskell but I don't know if they were higher than Bitwise operators. The reference list for the default groups can be found [here](https://github.com/apple/swift-evolution/blob/master/proposals/0077-operator-precedence.md#standard-library-changes). (The particular commit that changes this is [here](https://github.com/mozilla-mobile/firefox-ios/commit/019ee40eb85a57d45ae029a88ac4c6df17485048) /cc @rnewman)
* Fixed some warnings that were lying around. I'm not entirely sure about the `CFURLCreateStringByAddingPercentEscape` though so feel free to dig into that  🤔.

Unfortuantely for me to run the unit tests, it needs Client to host the unit test bundle which doesn't compile. We'll need to finish migrating that before we can run our test bundles 😞

farhanpatel commented:
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
AuthenticationKeychainInfo.swift | 9 | Colons should be next to the identifier when specifying a type. 
NSURLExtensions.swift | 8 | Colons should be next to the identifier when specifying a type. 
NSURLExtensions.swift | 21 | Colons should be next to the identifier when specifying a type. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

farhanpatel commented on Shared/Extensions/NSFileManagerExtensions.swift:
```diff
@@ -10,60 +10,60 @@ import Foundation
+public extension FileManager {
 
-    private func directoryEnumeratorForURL(url: NSURL) throws -> NSDirectoryEnumerator {
+    fileprivate func directoryEnumeratorForURL(_ url: URL) throws -> FileManager.DirectoryEnumerator {
```
I dont know what we wanna do with cases like this. Should we rename it correctly or just aim to get it work for now. `directoryEnumeratorForURL(_ url: URL)` could be `directoryEnumerator(forURL url: URL)`

farhanpatel commented on Shared/Extensions/NSFileManagerExtensions.swift:
```diff
@@ -10,60 +10,60 @@ import Foundation
     }
 
-    func allocatedSizeOfDirectoryAtURL(url: NSURL, forFilesPrefixedWith prefix: String, isLargerThanBytes threshold: Int64) throws -> Bool {
+    func allocatedSizeOfDirectoryAtURL(_ url: URL, forFilesPrefixedWith prefix: String, isLargerThanBytes threshold: Int64) throws -> Bool {
```
could be `allocatedSizeOfDirectory(at url: URL)`

fluffyemily replied:
I'd say `allocatedSizeOfDirectory(atURL url: NSURL, ...)`

farhanpatel commented on Client/Frontend/Strings.swift:
```diff
@@ -110,7 +110,7 @@ extension Strings {
     public static let FirefoxSyncTroubleshootTitle = NSLocalizedString("Settings.TroubleShootSync.Title", value: "Troubleshoot", comment: "Title of link to help page to find out how to solve Sync issues")
 
-    public static func localizedStringForSyncComponent(componentName: String) -> String? {
+    public static func localizedStringForSyncComponent(_ componentName: String) -> String? {
```
`localizedStringForSync(component name: String)`

farhanpatel commented on Shared/AuthenticationKeychainInfo.swift:
```diff
@@ -35,68 +48,68 @@ public extension KeychainWrapper {
-    private(set) public var failedAttempts: Int
-    public var useTouchID: Bool
+open class AuthenticationKeychainInfo: NSObject, NSCoding {
+    fileprivate(set) open var lastPasscodeValidationInterval: TimeInterval?
```
So these are fileprivate (accessible within the file) but then also open? (accessible but not overridible from anywhere) 

This is not quite right is it?

sleroux replied:
`private` here would be better since we have no intention of using this outside this lexical scope.

fluffyemily commented on Shared/Extensions/UIImageExtensions.swift:
```diff
@@ -21,44 +21,44 @@ extension UIImage {
-        var header = [UInt8](count: 3, repeatedValue: 0)
-        data.getBytes(&header, length: 3 * sizeof(UInt8))
+        var header = [UInt8](repeating: 0, count: 3)
+        (data as NSData).getBytes(&header, length: 3 * MemoryLayout<UInt8>.size)
```
apparently, according to StackOverflow anyway, the way to do this in Swift 3.0 is

`data.copyBytes(to: &header, count: 3 * MemoryLayout<UInt8>.size)`

fluffyemily commented on Shared/Extensions/UIImageExtensions.swift:
```diff
@@ -21,44 +21,44 @@ extension UIImage {
+        let rect = CGRect(origin: CGPoint.zero, size: size)
         color.setFill()
-        CGContextFillRect(context!, rect)
+        context!.fill(rect)
```
nit: I know `context` should always exist, but should we wrap this in a `guard` or `if var` instead of force unwrapping?

sleroux replied:
The force unwrap is awful but I'm a bit wary of making this change now since I'd ideally like to pass the optional through but existing callers expect an image. I'm not sure if a blank image is a good enough default state.

fluffyemily commented on Shared/Extensions/UIImageExtensions.swift:
```diff
@@ -21,44 +21,44 @@ extension UIImage {
 
-    public static func templateImageNamed(name: String) -> UIImage? {
-        return UIImage(named: name)?.imageWithRenderingMode(.AlwaysTemplate)
+    public static func templateImageNamed(_ name: String) -> UIImage? {
```
nit: `templateImage(named name: String)`

fluffyemily commented on Shared/Functions.swift:
```diff
@@ -3,27 +3,31 @@
 // Pipelining.
-infix operator |> { associativity left }
-public func |> <T, U>(x: T, f: T -> U) -> U {
+precedencegroup PipelinePrecedence {
```
This shit makes me so glad you did this and not me!

fluffyemily commented on Shared/Logger.swift:
```diff
@@ -50,25 +50,25 @@ public extension Logger {
     }
 
-    static private func fileLoggerWithName(name: String) -> XCGLogger {
+    static private func fileLoggerWithName(_ name: String) -> XCGLogger {
```
nit `fileLogger(withName name: String)`

fluffyemily commented on Shared/Prefs.swift:
```diff
@@ -0,0 +1,177 @@
+import Foundation
+
+public struct PrefsKeys {
+    public static let KeyLastRemoteTabSyncTime = "lastRemoteTabSyncTime"
```
nit: I'm pretty sure that Swift 3.0 style guide recommends lowerCamelCase constants https://github.com/raywenderlich/swift-style-guide#constants

fluffyemily commented on Shared/RollingFileLogger.swift:
```diff
@@ -6,16 +6,16 @@ import Foundation
 
-    private static let DateFormatter: NSDateFormatter = {
-        let formatter = NSDateFormatter()
+    fileprivate static let DateFormatter: Foundation.DateFormatter = {
```
Can we refer to `DateFormatter` without the explicit call to `Foundation`? If not, perhaps importing `Foundation` at the top is required?

fluffyemily commented on Shared/SupportUtils.swift:
```diff
@@ -11,11 +11,11 @@ public struct SupportUtils {
-    public static func URLForTopic(topic: String) -> NSURL? {
-        guard let escapedTopic = topic.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLPathAllowedCharacterSet()),
-                languageIdentifier = NSLocale.preferredLanguages().first else {
+    public static func URLForTopic(_ topic: String) -> URL? {
```
nit `url(forTopic topic: String)`

fluffyemily commented on ThirdParty/json.swift:
```diff
@@ -7,27 +7,27 @@
+    fileprivate let _value:AnyObject
     /// unwraps the JSON object
-    public class func unwrap(obj:AnyObject) -> AnyObject {
+    open class func unwrap(_ obj:AnyObject) -> AnyObject {
```
This could be `Any` instead of `AnyObject`

fluffyemily commented on ThirdParty/json.swift:
```diff
@@ -58,33 +58,33 @@ extension JSON {
+    public convenience init(nsurl:URL) {
+        var enc:String.Encoding = String.Encoding.utf8
         do {
             let str = try String(NSString(
```
no need to create `NSString` here. Can use `String(contentsOf: nsurl, encoding: enc)` instead

fluffyemily commented on ThirdParty/json.swift:
```diff
@@ -58,33 +58,33 @@ extension JSON {
     /// constructs JSON object from the content of URL
     public convenience init(url:String) {
-        if let nsurl = NSURL(string:url) as NSURL? {
+        if let nsurl = URL(string:url) as URL? {
```
no need to cast to `URL?` here. The result of `init(string:)` is already optional

sleroux commented on Shared/Extensions/NSFileManagerExtensions.swift:
```diff
@@ -79,27 +79,26 @@ public extension NSFileManager {
      - throws: Error reading/operating on disk.
      */
-    func getAllocatedSizeOfDirectoryAtURL(url: NSURL, forFilesPrefixedWith prefix: String) throws -> Int64 {
+    func getAllocatedSizeOfDirectoryAtURL(_ url: URL, forFilesPrefixedWith prefix: String) throws -> Int64 {
```
Yup that's definitely wonky

sleroux commented on Shared/Extensions/NSFileManagerExtensions.swift:
```diff
@@ -10,60 +10,60 @@ import Foundation
         }
 
-        return (url as? NSURL)?.allocatedFileSize() ?? 0
+        return (url as? URL)?.allocatedFileSize() ?? 0
```
Yup that's definitely wonky

sleroux commented:
Addressed all of the non-name change comments and cleaned up Data/json.swift a bit. I've filed a follow up issue here for renaming https://bugzilla.mozilla.org/show_bug.cgi?id=1334131.

fluffyemily commented:
The changes to JSON.swift make things so much cleaner. But it does mean I will have to revisit my Account migration to reflect these changes. Worth it though, I think.
